shell/key-bindings.bash

```diff
@@ -126,11 +133,11 @@ else
   bind -m emacs-standard -x '"\C-r": __fzf_history__'
   bind -m vi-command -x '"\C-r": __fzf_history__'
   bind -m vi-insert -x '"\C-r": __fzf_history__'
-fi
 
-# ALT-C - cd into the selected directory
-bind -m emacs-standard '"\ec": " \C-b\C-k \C-u`__fzf_cd__`\e\C-e\er\C-m\C-y\C-h\e \C-y\ey\C-x\C-x\C-d"'
-bind -m vi-command '"\ec": "\C-z\ec\C-z"'
-bind -m vi-insert '"\ec": "\C-z\ec\C-z"'
```

Comment:
We decided to keep the pre-bash4 implementation for ALT-C because of a critical UX issue. See https://github.com/junegunn/fzf/pull/2674.

Also, your version doesn't work and the test cases are failing. Have you tested it?

Reply:
Oops... I haven't noticed that it doesn't change the prompt, and was since then working on other branches.

I think I may have a solution for the issue, namely doing something like:
```sh
__fzf_cd__() {
  …
  dir=$(set +o pipefail; eval "$cmd" | FZF_DEFAULT_OPTS="$opts" $(__fzfcmd)) && builtin cd "$dir"
  bind '"\e[0n": "\C-j"' 2> /dev/null
  printf '\e[5n'
}
```
of course that's just a sketch and would require to restore the original `bind`.

Do you think that could be a way to go, or are there any issues with it, that I don't see right now?

But speaking of which, AFAICS, all he current code that uses the `\e[0n`/`\e[5n` trick sets the bind only once. Wouldn't it be generally better to set it again every time? I mean in case there are other scripts loaded, which use it too?


As for the failing tests... could be that some of the tests were really failing because of that bug, but I've had generally failing/flaky tests, even on `master` (see https://github.com/junegunn/fzf/pull/3449#issuecomment-1734544997).

Reply:
With the above implementation idea, a directory change looks like this:
```
calestyo@host:~$ 
calestyo@host:~/dir$ 
```
That is, the line where `Alt-C` was used is empty and the one afterwards has magically a new CWD.

I'd prefer if it would print `cd dir` in that line, instead of being empty, so that one can follow the transcript more easily.

But so far I'm only able to get it printed before the prompt :-/

Reply:
The above has of course still the problem that `C-j` may be bound to something else :-(

Reply:
Ooops... I hadn't noticed that this didn’t change the CWD, and worked on other branches since then and not looked back.

As for the tests, I've seen failures and flaky tests even when working on `master`, see https://github.com/junegunn/fzf/pull/3449#issuecomment-1734544997.
Not sure what I'm doing wrong there.

Okay, using `bind -x` seems indeed not so easy. `redraw-current-line` only re-draws the line, but does not re-evaluate `PS1`, so while the CWD has in fact changed, the old one is still shown in the prompt.<br>
Anything that causes enter to be printed wouldn’t be much better than the current solution (and cause the current line to be executed.

While searching for solutions, I stumbled over `${PS1@P}` - not sure if that could somehow be used to re-draw the prompt manually.<br>
Some people suggested sending `SIGWINCH` to the shell, but that didn’t work for me.

https://superuser.com/a/1662149 worked for me, but it would also require us override more bindings, and needs more shell vars. Same with https://stackoverflow.com/a/40426702.

The current implementation still seems fragile to me (in case anyone has chanced the meaning of any of those keybindings, which I for example actually have for `C-h`).<br>
Maybe we could somehow save the current keybindings, remove them (to be sure defaults are used), and restore them afters wards?

Reply:
@junegunn, I've removed all commits except the minor fa6b43fa9d3d7435209c3b773945980d764e18c5.

As for switching `Alt-C` to use `bind -x`:<br>
I still think there is some value in doing this, namely to get it working for any systems where the user has reassigned the keyseqs you use in:
https://github.com/junegunn/fzf/blob/4fdc08295b90485b7ee30a35d77177699c070113/shell/key-bindings.bash#L132-L134
and also to avoid having to set up helper bindings as in:
https://github.com/junegunn/fzf/blob/4fdc08295b90485b7ee30a35d77177699c070113/shell/key-bindings.bash#L102-L107
which may overwrite the user’s own bindings.

I think I may have found a working solution, at least for bash versions that already have `bind -x`.<br>
It's not yet perfect and I'm not enough of a bash/readline expert to determine whether it really works properly and in all cases, so I wrote mails to [help-bash](https://lists.gnu.org/mailman/listinfo/help-bash), asking for their advice:
- https://lists.gnu.org/archive/html/help-bash/2023-10/msg00000.html
- https://lists.gnu.org/archive/html/help-bash/2023-10/msg00001.html
if you're curious, you can already look there, how I think we could perhaps do it.

At least this time I actually did some testing O:-) ... and it looked good, except for the open points mentioned in the mail to help-bash.

I think we should handle any possible improvements to `Alt-C` in a separate issue, so if you're fine with the remaining minor commit of this PR, please merge it.<br>
At least the tests seem to ran through this time.